drivers/can/can_loopback.c

```diff
@@ -0,0 +1,225 @@
+ */
+
+#include <string.h>
+#include <kernel.h>
+#include <stdbool.h>
+#include <can.h>
+#include "can_loopback.h"
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(can_driver, CONFIG_CAN_LOG_LEVEL);
+
+static void dispatch_frame(const struct zcan_frame *frame,
+			   struct can_loopback_filter *filter)
+{
+	struct zcan_frame frame_tmp = *frame;
+
+	filter->rx_cb(&frame_tmp, filter->cb_arg);
+}
+
+static inline int check_filter_match(const struct zcan_frame *frame,
+				     const struct zcan_filter *filter)
+{
+	u32_t id, mask, frame_id;
+
+	frame_id = frame->id_type == CAN_STANDARD_IDENTIFIER ?
+			frame->std_id : frame->ext_id;
+	id = filter->id_type == CAN_STANDARD_IDENTIFIER ?
+			filter->std_id : filter->ext_id;
+	mask = filter->id_type == CAN_STANDARD_IDENTIFIER ?
+			filter->std_id_mask : filter->ext_id_mask;
+
+	return ((id & mask) == (frame_id & mask));
+}
+
+int can_loopback_send(struct device *dev, const struct zcan_frame *frame,
+		      s32_t timeout, can_tx_callback_t callback,
+		      void *callback_arg)
+{
+	struct can_loopback_data *data = DEV_DATA(dev);
+	int i;
+	struct can_loopback_filter *filter;
+
+	LOG_DBG("Sending %d bytes on %s. "
+		    "Id: 0x%x, "
+		    "ID type: %s, "
+		    "Remote Frame: %s"
+		    , frame->dlc, dev->config->name
+		    , frame->id_type == CAN_STANDARD_IDENTIFIER ?
+				frame->std_id : frame->ext_id
+		    , frame->id_type == CAN_STANDARD_IDENTIFIER ?
+		    "standard" : "extended"
+		    , frame->rtr == CAN_DATAFRAME ? "no" : "yes");
+
+	if (!data->loopback) {
+
+		return 0;
+	}
+
+	k_mutex_lock(&data->mtx, timeout);
+	for (i = 0; i < CONFIG_CAN_MAX_FILTER; i++) {
+		filter = &data->filters[i];
+		if (filter->rx_cb) {
+			if (check_filter_match(frame, &filter->filter)) {
+				dispatch_frame(frame, filter);
+			}
+		}
+	}
+
+	k_mutex_unlock(&data->mtx);
+
+	if (callback) {
+		callback(CAN_TX_OK, callback_arg);
+	}
+
+	return 0;
+}
+
+
+static inline int get_free_filter(struct can_loopback_filter *filters)
+{
+	int i;
+
+	for (i = 0; i < CONFIG_CAN_MAX_FILTER; i++) {
+		if (filters[i].rx_cb == NULL) {
+			return i;
+		}
+	}
+
+	return -1;
+}
+
+int can_loopback_attach_isr(struct device *dev, can_rx_callback_t isr,
+			    void *cb_arg,
+			    const struct zcan_filter *filter)
+{
+	struct can_loopback_data *data = DEV_DATA(dev);
+	struct can_loopback_filter *loopback_filter;
+	int filter_id;
+
+	LOG_DBG("Setting filter ID: 0x%x, mask: 0x%x", filter->ext_id,
+		    filter->ext_id_mask);
+	LOG_DBG("Filter type: %s ID %s mask",
+		    filter->id_type == CAN_STANDARD_IDENTIFIER ?
+			"standard" : "extended",
+		     ((filter->id_type && (filter->std_id_mask == CAN_STD_ID_MASK)) ||
+		     (!filter->id_type && (filter->ext_id_mask == CAN_EXT_ID_MASK))) ?
+			"with" : "without");
+
+	k_mutex_lock(&data->mtx, K_FOREVER);
+	filter_id = get_free_filter(data->filters);
+
+	if (filter_id < 0) {
+		LOG_ERR("No free filter left");
+		k_mutex_unlock(&data->mtx);
+		return CAN_NO_FREE_FILTER;
```

Comment:
I would suggest using a standard error number. Something like -EBUSY. I am guessing there is no negative filter ID right?

Reply:
Right. But the CAN error numbers are defined along with the API to get a more precise error message.